src/doc/trpl/error-handling.md

```diff
@@ -1,308 +1,2147 @@
+(like [`ErrorKind`](http://doc.rust-lang.org/std/io/enum.ErrorKind.html))
+or keep it hidden
+(like
+[`ParseIntError`](http://doc.rust-lang.org/std/num/struct.ParseIntError.html)).
+Regardless of how you do it, it's usually good practice to at least provide
+some information about the error beyond just its `String` representation. But
+certainly, this will vary depending on use cases.
+
+At a minimum, you should probably implement the
+[`Error`](http://doc.rust-lang.org/std/error/trait.Error.html)
+trait. This will give users of your library some minimum flexibility for
+[composing errors](#the-real-try-macro). Implementing the `Error` trait also
+means that users are guaranteed the ability to obtain a string representation
+of an error (because it requires impls for both `fmt::Debug` and
+`fmt::Display`).
+
+Beyond that, it can also be useful to provide implementations of `From` on your
+error types. This allows you (the library author) and your users to
+[compose more detailed errors](#composing-custom-error-types). For example,
+[`csv::Error`](http://burntsushi.net/rustdoc/csv/enum.Error.html)
+provides `From` impls for both `io::Error` and `byteorder::Error`.
+
+Finally, depending on your tastes, you may also want to define a
+[`Result` type alias](#the-result-type-alias-idiom), particularly if your
+library defines a single error type. This is used in the standard library
+for [`io::Result`](http://doc.rust-lang.org/std/io/type.Result.html)
+and [`fmt::Result`](http://doc.rust-lang.org/std/fmt/type.Result.html).
+
+
+## Case study: A program to read population data
+
+This chapter was long, and depending on your background, it might be rather
+dense. While there is plenty of example code to go along with the prose, most
+of it was specifically designed to be pedagogical. While I'm not quite smart
+enough to craft pedagogical examples that are also *not* toy examples, I
+certainly can write about a case study.
+
+For this, I'd like to build up a command line program that lets you
+query world population data. The objective is simple: you give it a location
+and it will tell you the population. Despite the simplicity, there is a lot
+that can go wrong!
+
+The data we'll be using comes from the
+[Data Science Toolkit](https://github.com/petewarden/dstkdata). I've prepared
+some data from it for this exercise. You can either grab the
+[world population data](http://burntsushi.net/stuff/worldcitiespop.csv.gz)
+(41MB gzip compressed, 145MB uncompressed) or just the
+[US population data](http://burntsushi.net/stuff/uscitiespop.csv.gz)
+(2.2MB gzip compressed, 7.2MB uncompressed).
+
+Up until now, I've kept the code limited to Rust's standard library. For a real
+task like this though, we'll want to at least use something to parse CSV data,
+parse the program arguments and decode that stuff into Rust types automatically. For that, we'll use the
+[`csv`](https://crates.io/crates/csv),
+[`docopt`](https://crates.i
```

Comment:
Yeah, I also think that this is a side effect weakness of the "everything in Cargo" policy, and one we don't have a great answer to. And I love docopt personally, but would feel bad because `clap` has a great following too...